cpp/include/cuvs/neighbors/ivf_list.hpp

```diff
@@ -0,0 +1,241 @@
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <raft/core/device_mdarray.hpp>
+#include <raft/core/error.hpp>
+#include <raft/core/host_mdarray.hpp>
+#include <raft/core/mdspan.hpp>
+#include <raft/core/resource/cuda_stream.hpp>
+#include <raft/core/resource/thrust_policy.hpp>
+#include <raft/core/resources.hpp>
+#include <raft/core/serialize.hpp>
+#include <raft/util/integer_utils.hpp>
+
+#include <thrust/fill.h>
+
+#include <atomic>
+#include <fstream>
+#include <limits>
+#include <memory>
+#include <type_traits>
+
+namespace cuvs::neighbors::ivf {
+
+/**
+ * Default value filled in the `indices` array.
+ * One may encounter it trying to access a record within a list that is outside of the
+ * `size` bound or whenever the list is allocated but not filled-in yet.
+ */
+template <typename IdxT>
+constexpr static IdxT kInvalidRecord =
+  (std::is_signed_v<IdxT> ? IdxT{0} : std::numeric_limits<IdxT>::max()) - 1;
+
+/** The data for a single IVF list. */
+template <template <typename, typename...> typename SpecT,
+          typename SizeT,
+          typename... SpecExtraArgs>
+struct list {
+  using size_type    = SizeT;
+  using spec_type    = SpecT<size_type, SpecExtraArgs...>;
+  using value_type   = typename spec_type::value_type;
+  using index_type   = typename spec_type::index_type;
+  using list_extents = typename spec_type::list_extents;
+
+  /** Possibly encoded data; it's layout is defined by `SpecT`. */
+  raft::device_mdarray<value_type, list_extents, raft::row_major> data;
+  /** Source indices. */
+  raft::device_mdarray<index_type, raft::extent_1d<size_type>, raft::row_major> indices;
+  /** The actual size of the content. */
+  std::atomic<size_type> size;
+
+  /** Allocate a new list capable of holding at least `n_rows` data records and indices. */
+  list(raft::resources const& res, const spec_type& spec, size_type n_rows)
+    : size{n_rows}, data{res}, indices{res}
+  {
+    auto capacity = raft::round_up_safe<SizeT>(n_rows, spec.align_max);
+    if (n_rows < spec.align_max) {
+      capacity = raft::bound_by_power_of_two<SizeT>(std::max<SizeT>(n_rows, spec.align_min));
+      capacity = std::min<SizeT>(capacity, spec.align_max);
+    }
+    try {
+      data    = raft::make_device_mdarray<value_type>(res, spec.make_list_extents(capacity));
+      indices = raft::make_device_vector<index_type, SizeT>(res, capacity);
+    } catch (std::bad_alloc& e) {
+      RAFT_FAIL(
+        "ivf::list: failed to allocate a big enough list to hold all data "
+        "(requested size: %zu records, selected capacity: %zu records). "
+        "Allocator exception: %s",
+        size_t(size),
+        size_t(capacity),
+        e.what());
+    }
+    // Fill the index buffer with a pre-defined marker for easier debugging
```

Comment:
Implementations should no longer be included in header files in the public APIs and the only APIs in `include/` should be interfaces that define the contract for users. Also, all interfaces exposed in `include/` should be comprehensively documented with usage examples.